[team4_pair2] step4 베스트 반찬 Tab UI 및 백엔드 캐시 구현

Tomadoing opened the pull request.

Commit:
User 인증 관련 로직 변경

- UserService의 Session 키를 사용하도록 변경
- 사용하지 않는 LoginUser Annotation 및 ArgumentResolver 제거
- import 최적화

Commit:
캐시 설정을 위한 의존성 추가

Commit:
베스트 카테고리 API 추가

- 베스트 카테고리 목록 및 관련 반찬 조회 API 추가
- 성능 향상을 위한 캐시 적용
- JSON 응답 포맷을 일관성을 위한 RestResponse 추가
- 기본 데이터 추가

Commit:
베스트 반찬 Tab UI 구현

- 페이지 로드시 동적으로 베스트 반찬 목록 렌더링
- 반찬 목록 클릭시 관련 반찬을 동적으로 보여주도록 변경

crongro commented on src/main/resources/static/js/category.js:
```diff
@@ -1,21 +1,80 @@
         .then(validateResponse)
         .then(({data}) => {
             const menu = $("#gnb .menu");
             menu.addEventListener('mouseover', showSubCategories);
```
addEventListener 하는 코드는 미리 할 수 있는거 같은데요? 
menu 엘리먼트가 존재하면  then안에서 하지말고, domcontentloaded 시점에서 별도로 하면 좋겠어요.

crongro commented on src/main/resources/static/js/category.js:
```diff
@@ -1,21 +1,80 @@
+        .then(validateResponse)
+        .then(({data}) => {
+            const eventMenu = $(".tab-btn-box");
+            eventMenu.addEventListener('click', handleEventBanchanClicked);
```
이것도 여기서 꼭 할필요 없음.
then을 통한 데이터 처리흐름에 엉뚱한 게 끼어들어가서 흐름을 끊는 느낌.

crongro commented on src/main/resources/static/js/category.js:
```diff
@@ -1,21 +1,80 @@
+            const eventMenu = $(".tab-btn-box");
+            eventMenu.addEventListener('click', handleEventBanchanClicked);
+            data.forEach(eventCategory => {
+                eventMenu.innerHTML += eventCategoryTemplate(eventCategory);
```
반복문안에서 innerHTML을 쓰면 연속적으로 DOM처리가 진행됨으로
한번만 DOM에 추가하는 방식으로 변경해야할 듯.

crongro commented on src/main/resources/static/js/category.js:
```diff
@@ -45,6 +104,61 @@ function categoryTemplate({id, name}, has_submenu) {
+                    <p class="txt">전국택배</p>
+                  </div>
+                  <div class="badge-wrapper">
+                    ${badges && badges.map(badgeTemplate).join('')}
```
map 구현 ! 짝짝짝

crongro reviewed (approved):
오 잘 했네요.
리뷰참고로 남깁니다.

hyukjin-lee commented on src/main/java/codesquad/banchan/domain/Banchan.java:
```diff
@@ -0,0 +1,42 @@
+    private String title;
+
+    @Column(nullable = false)
+    private String subTitle;
```
hey guys! subTitle is not two word. you need to merge this words. Horrible~

minsungryu replied:
o..okay; 😭

minsungryu commented:
@nigayo 좋은 피드백 감사합니다! 얼른 개선해볼게요

Commit:
F/E 피드백 반영

minsungryu committed:
Banchan class 변수명 변경

- 테스트 데이터 컬럼 수정

minsungryu committed:
F/E 리팩토링

- 프로모션 영역 트랜지션 속도 단축
- 프로모션 불필요한 z-index 제거(sub-menu에 영향)
- 프로모션 dot list 동적으로 변경
- Element.closest 사용하여 탐색
- 로그아웃 addEventListener 예외 처리

javajigi commented on src/main/java/codesquad/bestcategory/web/ApiBestCategoryController.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    @GetMapping("/{id}/banchans")
+    @Cacheable(value = "bestCategoryBanchan")
```
Cache 데이터에 변경이 발생할 경우 Cache를 무효화하기 위해 CacheEvict이라는 설정을 사용할 수 있어요.
이와 관련한 학습하고 추후 적용해 보세요.

minsungryu replied:
5단계 반찬 관리 기능 만들때 적용해 보겠습니다!

javajigi reviewed (approved):
Cache 관련 피드백 하나 남겼어요.
